elements/lisk-p2p/src/peer/base.ts

```diff
@@ -173,21 +184,53 @@ export class Peer extends EventEmitter {
 	) => void;
 	protected _socket: SCServerSocketUpdated | SCClientSocket | undefined;
 
-	public constructor(peerInfo: P2PPeerInfo, peerConfig?: PeerConfig) {
+	public constructor(peerInfo: P2PPeerInfo, peerConfig: PeerConfig) {
 		super();
 		this._peerInfo = peerInfo;
-		this._peerConfig = peerConfig ? peerConfig : {};
+		this._peerConfig = peerConfig;
 		this._ipAddress = peerInfo.ipAddress;
 		this._wsPort = peerInfo.wsPort;
 		this._id = constructPeerId(this._ipAddress, this._wsPort);
 		this._height = peerInfo.height ? (peerInfo.height as number) : 0;
 		this._reputation = DEFAULT_REPUTATION_SCORE;
 		this._latency = 0;
 		this._connectTime = Date.now();
-		this._callCounter = new Map();
+		this._rpcCounter = new Map();
+		this._rpcRates = new Map();
+		this._messageCounter = new Map();
+		this._messageRates = new Map();
+		this._wsMessageCount = 0;
+		this._wsMessageRate = 0;
+		this._rateInterval = this._peerConfig.rateCalculationInterval;
 		this._counterResetInterval = setInterval(() => {
-			this._callCounter = new Map();
-		}, DEFAULT_RATE_INTERVAL);
+			this._wsMessageRate =
+				(this._wsMessageCount * RATE_NORMALIZATION_FACTOR) / this._rateInterval;
+			this._wsMessageCount = 0;
+
+			if (this._wsMessageRate > this._peerConfig.wsMaxMessageRate) {
+				this.disconnect(FORBIDDEN_CONNECTION, FORBIDDEN_CONNECTION_REASON);
+
+				return;
+			}
+
+			this._rpcRates = new Map(
+				[...this._rpcCounter.entries()].map(entry => {
+					const rate = entry[1] / this._rateInterval;
+
+					return [entry[0], rate] as any;
```

Comment:
isn't `entry[0]` is string type and we can use it as key for `rpcRates` type as `Map<string, number>`?

Reply:
we don't need to typecast as `any`, it doesn't give me warning

Reply:
It gives me warnings in vscode if I don't put `any`.